### PR TITLE
Elimination mode [#410]

### LINKED ISSUE #410: Add elimination mode
Limit of how many lives one has, and if the user runs out, they're eliminated. Should work in single-player as well

### PR DESCRIPTION
**Pull request by taahamahdi**
Closes #410.

### CODE REVIEW

File: src/commands/game_commands/play.ts [hunk 1]
```diff
@@ -1,15 +1,23 @@
...
 } from "../../helpers/discord_utils";
 import { getGuildPreference } from "../../helpers/game_utils";
+import { bold } from "../../helpers/utils";
 import BaseCommand, { CommandArgs } from "../base_command";
 import _logger from "../../logger";
 
 const logger = _logger("play");
 
+export enum GameType {
+    ELIMINATION = "elimination",
```
Comment (Brainicism) [inline src/commands/game_commands/play.ts:14]: Add GameType for default

File: src/models/elimination_scoreboard.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        Object.values(this.players).forEach((player) => {
+            player.decrementLives();
+        });
+    }
+
+    allPlayersEliminated() {
+        return Object.values(this.players).every((player) => player.isEliminated());
+    }
+
+    onePlayerLeft(): boolean {
```
Comment (Brainicism) [inline src/models/elimination_scoreboard.ts:75]: This name is misleading in the case of one player

File: src/models/game_session.ts [hunk 3]
```diff
@@ -291,8 +310,20 @@ export default class GameSession {
         this.guessTimeoutFunc = setTimeout(async () => {
             if (this.finished) return;
             logger.info(`${getDebugContext(message)} | Song finished without being guessed, timer of: ${time} seconds.`);
+            if (this.eliminationMode) {
+                const eliminationScoreboard = this.scoreboard as EliminationScoreboard;
+                eliminationScoreboard.decrementAllLives();
+            }
             sendSongMessage(message, this.scoreboard, this.gameRound, true);
             this.endRound(false);
+            if (this.eliminationMode) {
```
Comment (Brainicism) [inline src/models/game_session.ts:319]: Combine these two if statements?

File: src/models/scoreboard.ts [hunk 4]
```diff
@@ -88,4 +84,17 @@ export default class Scoreboard {
...
+    awardPoint(winnerTag: string, winnerID: string, avatarURL: string, pointsEarned: number) {
+        if (!this.players[winnerID]) {
+            this.players[winnerID] = new Player(winnerTag, winnerID, avatarURL, pointsEarned);
+        } else {
+            this.players[winnerID].incrementScore(pointsEarned);
+        }
+    }
+
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    setPlayers(players: { [userID: number]: {tag: string, avatar: string} }) {
```
Comment (Brainicism) [inline src/models/scoreboard.ts:97]: Remove this, cast scoreboard to EliminationScoreboard when needed

File: src/commands/game_commands/play.ts [hunk 5]
```diff
@@ -23,22 +31,95 @@ export default class PlayCommand implements BaseCommand {
...
+                const gameOwner = message.author;
+                const lives = parsedMessage.components.length > 1 ? parseInt(parsedMessage.components[1], 10) : DEFAULT_LIVES;
+                const gameSession = new GameSession(textChannel, voiceChannel, gameOwner, eliminationMode, lives);
                 gameSessions[message.guildID] = gameSession;
-                await sendInfoMessage(message, `Game starting in #${textChannel.name} in 🔊 ${voiceChannel.name}`, "Listen to the song and type your guess!");
-                logger.info(`${getDebugContext(message)} | Game session starting`);
-                gameSession.startRound(guildPreference, message);
+
+                const lifeOrLives = lives === 1 ? "life" : "lives";
+                const joinReaction = "✋";
```
Comment (Brainicism) [inline src/commands/game_commands/play.ts:41]: Move these constants to the top

File: src/commands/game_commands/play.ts [hunk 6]
```diff
@@ -23,22 +31,95 @@ export default class PlayCommand implements BaseCommand {
...
+                    startMessage.addReaction(joinReaction);
+                    startMessage.addReaction(startReaction);
+                    state.client.on("messageReactionAdd", (msg, emoji, reactor) => {
+                        if (msg.id !== startMessage.id || emoji.name !== startReaction || reactor !== gameOwner.id) {
+                            return;
+                        }
+                        const participants: { [userID: string]: {tag: string, avatar: string} } = {};
+
+                        // When startReaction pressed, get users who reacted to joinReaction and set as participants
+                        Promise.resolve(state.client.getMessageReaction(textChannel.id, msg.id, joinReaction))
```
Comment (Brainicism) [inline src/commands/game_commands/play.ts:56]: Can't you just `await` this

File: src/commands/game_commands/play.ts [hunk 7]
```diff
@@ -23,22 +31,95 @@ export default class PlayCommand implements BaseCommand {
...
+                            .then((reactors) => {
+                                reactors.forEach((player) => {
```
Comment (Brainicism) [inline src/commands/game_commands/play.ts:58]: .reduce() is probably a cleaner solution

File: src/commands/game_commands/play.ts [hunk 8]
```diff
@@ -23,22 +31,95 @@ export default class PlayCommand implements BaseCommand {
...
+                                    if (player.id === state.client.user.id) {
+                                        return;
+                                    }
+                                    participants[player.id] = { tag: getUserIdentifier(player), avatar: player.avatarURL };
+                                });
+                                if (Object.keys(participants).length === 0) return;
```
Comment (Brainicism) [inline src/commands/game_commands/play.ts:64]: Should there be a warning message in this case?

File: src/commands/game_commands/play.ts [hunk 9]
```diff
@@ -23,22 +31,95 @@ export default class PlayCommand implements BaseCommand {
...
+                example: "`!play`",
+                explanation: "Start a classic game of KMQ (type in your guess first to get a point)",
+            },
+            {
+                example: "`!play elimination 5`",
+                explanation: "Start an elimination game of KMQ where each player starts with 5 lives.",
+            },
+            {
+                example: "`!play elimination`",
+                explanation: `Start an elimination game of KMQ where each player starts with ${DEFAULT_LIVES} lives.`,
```
Comment (Brainicism) [inline src/commands/game_commands/play.ts:121]: Put code block around default lives value

File: src/helpers/discord_utils.ts [hunk 10]
```diff
@@ -219,7 +218,7 @@ export async function sendScoreboardMessage(message: Eris.Message<Eris.GuildText
         title: "**Scoreboard**",
         fields: winnersFieldSubset,
         footer: {
-            text: `Your score is ${gameSession.scoreboard.getPlayerScore(message.author.id)}.`,
+            text: gameSession.eliminationMode ? "" : `Your score is ${gameSession.scoreboard.getPlayerScore(message.author.id)}.`,
```
Comment (Brainicism) [inline src/helpers/discord_utils.ts:221]: Should this tell them how many lives they have?

File: src/models/elimination_scoreboard.ts [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+        super();
+        this.startingLives = lives;
+    }
+
+    getScoreboardEmbedFields(): Array<{ name: string, value: string, inline: boolean }> {
+        return Object.values(this.players)
+            .sort((a, b) => b.getLives() - a.getLives())
+            .map((x) => {
+                const heartEmoji = "❤️";
+                const lives = x.getLives() <= 5 ? heartEmoji.repeat(x.getLives()) : `${heartEmoji} x ${x.getLives()}`;
```
Comment (Brainicism) [inline src/models/elimination_scoreboard.ts:22]: Showing x number of hearts will add too much clutter with large number of players

**Review by Brainicism — CHANGES_REQUESTED** [review]

File: src/commands/game_commands/play.ts (hunk 1 above)
Comment (taahamahdi) [reply src/commands/game_commands/play.ts:14]: In the validations it'll show that the user can enter either elimination or default if I make this change. Should I allow them to do so?

File: src/commands/game_commands/play.ts (hunk 8 above)
Comment (taahamahdi) [reply src/commands/game_commands/play.ts:64]: As it is now, it'll ignore the reaction and let the user retry as many times as needed before someone has joined as a participant. Spamming this reaction could lead to multiple warning messages being sent, which isn't ideal. However, I could limit this to only once if you'd prefer a warning message.

File: src/models/elimination_scoreboard.ts (hunk 11 above)
Comment (taahamahdi) [reply src/models/elimination_scoreboard.ts:22]: I'll replace it with the number of lives

File: src/commands/game_commands/play.ts (hunk 6 above)
Comment (taahamahdi) [reply src/commands/game_commands/play.ts:56]: I needed to add async to the anonymous function it's in, now that it's there it should work. Thanks

File: src/commands/game_commands/play.ts [hunk 12]
```diff
@@ -1,15 +1,26 @@
...
 } from "../../helpers/discord_utils";
 import { getGuildPreference } from "../../helpers/game_utils";
+import { bold } from "../../helpers/utils";
 import BaseCommand, { CommandArgs } from "../base_command";
 import _logger from "../../logger";
 
 const logger = _logger("play");
 
+export enum GameType {
+    DEFAULT = "default",
```
Comment (Brainicism) [inline src/commands/game_commands/play.ts]: default => classic

File: src/commands/game_commands/play.ts [hunk 13]
```diff
@@ -23,22 +34,91 @@ export default class PlayCommand implements BaseCommand {
...
+                const eliminationMode = parsedMessage.components.length >= 1 && parsedMessage.components[0].toLowerCase() === "elimination";
+                const gameOwner = message.author;
+                const lives = parsedMessage.components.length > 1 ? parseInt(parsedMessage.components[1], 10) : DEFAULT_LIVES;
+                const gameSession = new GameSession(textChannel, voiceChannel, gameOwner, eliminationMode, lives);
                 gameSessions[message.guildID] = gameSession;
-                await sendInfoMessage(message, `Game starting in #${textChannel.name} in 🔊 ${voiceChannel.name}`, "Listen to the song and type your guess!");
-                logger.info(`${getDebugContext(message)} | Game session starting`);
-                gameSession.startRound(guildPreference, message);
+
+                const lifeOrLives = lives === 1 ? "life" : "lives";
```
Comment (Brainicism) [inline src/commands/game_commands/play.ts:43]: Easier just to replace with "live(s)"

**Review by Brainicism — CHANGES_REQUESTED** [review]

File: src/commands/game_commands/play.ts (hunk 13 above)
Comment (taahamahdi) [reply src/commands/game_commands/play.ts:43]: I think I'll just keep it as lives, just realized the singular version (live) doesn't make sense

**Pull request closed without merging**